tracing/tracing.go

```diff
@@ -0,0 +1,82 @@
+package tracing
+
+import "context"
+
+type contextKey string
+
+const (
+	traceContextKey contextKey = "trace"
+)
+
+// TracerFromContext returns a tracer from the context. If no tracer is found,
+// an empty tracer is returned.
+func TracerFromContext(ctx context.Context) Tracer {
+	value := ctx.Value(traceContextKey)
+	if value == nil {
+		return noopTracer{}
+	}
+	tracer, ok := value.(Tracer)
+	if !ok {
+		return noopTracer{}
+	}
+	return tracer
+}
+
+// WithTracer returns a new context with the given tracer.
+func WithTracer(ctx context.Context, tracer Tracer) context.Context {
+	if tracer == nil {
+		tracer = noopTracer{}
+	}
+	return context.WithValue(ctx, traceContextKey, tracer)
+}
+
+// Start returns a new context with the given trace.
+func Start(ctx context.Context, name, query string) (context.Context, Span) {
+	// Tracer is always guaranteed to be returned here. If there is no tracer
+	// available it will return a noop tracer.
+	tracer := TracerFromContext(ctx)
+	return tracer.Start(ctx, name, query)
+}
+
+// Tracer is the interface that all tracers must implement.
+type Tracer interface {
+	// Start creates a span and a context.Context containing the newly-created
+	// span.
+	//
+	// If the context.Context provided in `ctx` contains a Span then the
+	// newly-created Span will be a child of that span, otherwise it will be a
+	// root span.
```

Comment:
How does an implementor of Tracer retrieve the "parent" span from the context? Presumably there's another string key that's used?

Reply:
The context that's passed in also contains the parent. If there is another trace started with the threaded context that has a parent id, then it will create a child span of the parent. Not all tracing libraries work this way, but [OTEL](https://opentelemetry.io/) does. We could add another string (or type) for the parent/child relationship, but I'd propose that we cross that bridge when we come to it.

Reply:
I think I understand now, the responsibility for somehow stashing the parent Span inside the context is entirely the implementor's -- I got confused because there was no declared protocol for retrieving the parent, but there doesn't need to be, each implementor can implement their own solution since they control all the relevant code.